### PR TITLE
Ublox GPS - Manual hot, warm and cold reset functionality

### PR DESCRIPTION
**Pull request by jesusBV20**
With this feature, we can force our Ublox GPS to make a cold reset when the AP is turned on. Thus we can simulate a cold start.

By adding the _reset_ field to the _gps_ubx_ structure I have also fixed the _gps_ubx_Reset_ function defined in _gps_ubx.h_.

### CODE REVIEW

File: sw/airborne/modules/gps/gps_ubx.h [hunk 1]
```diff
@@ -43,6 +43,10 @@
 #define PRIMARY_GPS GPS_UBX
 #endif
 
+#ifndef GPS_UBX_COLDSTART
+#define GPS_UBX_COLDSTART false
+#endif
+
```
Comment (gautierhattenberger) [inline sw/airborne/modules/gps/gps_ubx.h:49]: this part can be moved to the .c file

File: sw/airborne/modules/gps/gps_ubx.h [hunk 2]
```diff
@@ -65,6 +69,7 @@ struct GpsUbx {
   uint8_t send_ck_a, send_ck_b;
   uint8_t error_cnt;
   uint8_t error_last;
+  uint16_t reset;
```
Comment (gautierhattenberger) [inline sw/airborne/modules/gps/gps_ubx.h:72]: would it be interesting to make a setting to ask a reset from the ground station ?

File: sw/airborne/modules/gps/gps_ubx.c [hunk 3]
```diff
@@ -104,6 +104,12 @@ void gps_ubx_init(void)
   gps_ubx.error_last = GPS_UBX_ERR_NONE;
 
   gps_ubx.state.comp_id = GPS_UBX_ID;
+  
+#if GPS_UBX_COLDSTART
```
Comment (gautierhattenberger) [inline sw/airborne/modules/gps/gps_ubx.c:108]: it should be documented in the module's documentation (xml file)

**Review by gautierhattenberger — CHANGES_REQUESTED** [review]

**Comment by jesusBV20** [comment]
Done :)

File: conf/modules/gps_ublox.xml [hunk 4]
```diff
@@ -5,10 +5,24 @@
...
+        - Coldstart (reset = 3) You can force a cold reset when the AP is turned on by defining GPS_UBX_COLDSTART as TRUE in your airframe.
     </description>
     <configure name="UBX_GPS_PORT" value="UARTx" description="UART where the GPS is connected to (UART1, UART2, etc"/>
     <configure name="UBX_GPS_BAUD" value="B38400" description="UART baud rate"/>
   </doc>
+
+  <settings>
+    <dl_settings>
+      <dl_settings name="gps_ublox">
+        <dl_setting MIN="0" MAX="3" STEP="1" module="gps/gps_ubx" VAR="gps_ubx.reset" shortname="reset"/>
```
Comment (fvantienen) [inline conf/modules/gps_ublox.xml:21]: Can you add the Hotstart, Warmstart and Coldstart as well here as `values`?

File: conf/modules/gps_ublox.xml [hunk 5]
```diff
@@ -5,10 +5,24 @@
...
       U-blox GPS (UART)
       Driver for u-blox GPS modules parsing the binary UBX protocol.
+      
+      The reset variable allows us to manually restart the u-blox GPS. There are three modes:
+        - Hotstart  (reset = 1)
+        - Warmstart (reset = 2)
+        - Coldstart (reset = 3) You can force a cold reset when the AP is turned on by defining GPS_UBX_COLDSTART as TRUE in your airframe.
     </description>
     <configure name="UBX_GPS_PORT" value="UARTx" description="UART where the GPS is connected to (UART1, UART2, etc"/>
     <configure name="UBX_GPS_BAUD" value="B38400" description="UART baud rate"/>
```
Comment (fvantienen) [inline conf/modules/gps_ublox.xml:15]: Add the documentation for the `GPS_UBX_COLDSTART` here.

File: sw/airborne/modules/gps/gps_ubx.c (hunk 3 above)
Comment (fvantienen) [reply sw/airborne/modules/gps/gps_ubx.c:113]: Maybe add it as a define like this:
```
#ifndef GPS_UBX_BOOTRESET
#define GPS_UBX_BOOTRESET 0
#endif
...
 gps_ubx.reset = GPS_UBX_BOOTRESET;
```

Makes the code easier to read.

**Review by fvantienen — CHANGES_REQUESTED** [review]

File: sw/airborne/modules/gps/gps_ubx.h (hunk 2 above)
Comment (fvantienen) [reply sw/airborne/modules/gps/gps_ubx.h:68]: This `uint16_t` could also just be an `uint8_t`?

**Comment by jesusBV20** [comment]
Thanks @gautierhattenberger and @fvantienen for your suggestions. Changes have been made!

File: sw/airborne/modules/gps/gps_ubx.c [hunk 6]
```diff
@@ -86,6 +86,11 @@ extern struct RtcmMan rtcm_man;
 #ifndef INJECT_BUFF_SIZE
 #define INJECT_BUFF_SIZE 1024 + 6
 #endif
+
+#ifndef GPS_UBX_BOOTRESET
+#define GPS_UBX_BOOTRESET 0
+#endif
+
```
Comment (gautierhattenberger) [inline sw/airborne/modules/gps/gps_ubx.c:93]: this part is wrongly placed inside the `USE_GPS_UBX_RTCM` section, you should move it either before or after.

**Review by gautierhattenberger — CHANGES_REQUESTED** [review]

File: sw/airborne/modules/gps/gps_ubx.c (hunk 6 above)
Comment (jesusBV20) [reply sw/airborne/modules/gps/gps_ubx.c:93]: Oh, sorry for the mistake. I'll fix it in a moment.

**Review by fvantienen — APPROVED** [review]

**Review by gautierhattenberger — APPROVED** [review]

**Pull request merged**